### ISSUE TITLE
factory_boy fails to connect to database in GitLab CI

### ISSUE THREAD

**Issue by ovlb** [FactoryBoy/factory_boy#881]
Hey there,

I have the suspicion that I’m missing something self-evident. But I can’t figure out what.

So, I have a Django project with FB installed. But somehow, FB depends on the default database, which is fine, locally. But in CI, this DB is missing during our test step. 

The first error message I got was 

> django.db.utils.OperationalError: FATAL:  database "projectname" does not exist

After some googling, I found #171 and https://github.com/FactoryBoy/factory_boy/issues/171#issuecomment-71597920. 

I implemented the suggestion, which at least changes the error message:

> django.db.utils.ProgrammingError: relation "projectname_relatedschema" does not exist
LINE 1: ...hema_it", "projectname_relatedschema"."order" FROM "projectname...

If I rename my local default database, the tests are failing with the first error message.

I’m a bit lost here, since I don’t find any documentation around database usage and also no blog posts or Stack Overflow threads dealing with this problem. 

Has anyone encountered this problem previously and knows what I might be missing/overlooking?

Thanks,
Oscar

**Comment by rbarrois** [comment]
@ovlb Sorry for taking this long to reply; in your test setup, is the testing database created, alongside its tables?
Can you provide some more lines around the actual exception?

The main possibilities I see are:
- You're trying to run the tests without a SQL database available; in that case, make sure you're using `MyFactory.build()` to avoid touching the database;
- Your code is instantiating some factories at module import time, before the test database has been populated; make sure that code is only called once the database has been setup;
- Some of your factories declarations performs SQL queries at import time; use a `factory.LazyFunction` or similar;
- You're running with several databases, and FactoryBoy sends queries to the wrong one; use the [`Params.database`](https://factoryboy.readthedocs.io/en/latest/orms.html#factory.django.DjangoOptions.database) parameter to point it to the right database.

**Comment by ovlb** [comment]
@rbarrois No worries. I managed to figure it out in the meantime.

We had indeed an import problem. There was a foreign key in the factory which we used as `field_name = ForeignModel.objects.get(id=3)`. We changed this to `field_name = factory.Iterator(ForeignModel.objects.filter(id=3))`

Out of interest, is a `LazyFunction` or `LazyAttribute` more appropriate than a non-iterating iterator?

Thanks for getting back to me and for factory boy. It made our tests soooo much better.

Oscar
